### DISCUSSION TITLE
Some remarks about "Heroes of Might and Magic 2 Stream" 2022 04 17

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#5478] [General]
Here are some remarks about this video:
https://www.youtube.com/watch?v=eJq7L3kO2Ng

Meeting screen.
Here you can see an experienced player using the big yellow arrows for troop transfers.
Note that the player has chosen to always keep a fast troop in the hero's army with one troop.
https://youtu.be/eJq7L3kO2Ng?t=1924
https://youtu.be/eJq7L3kO2Ng?t=2365
https://youtu.be/eJq7L3kO2Ng?t=15068
https://youtu.be/eJq7L3kO2Ng?t=15375 (He comments on the transfer function.)
https://youtu.be/eJq7L3kO2Ng?t=15494 (He gave up.)
https://youtu.be/eJq7L3kO2Ng?t=15543 (One more try.)
https://youtu.be/eJq7L3kO2Ng?t=15762 (One more try.)
https://youtu.be/eJq7L3kO2Ng?t=15810 (He gave up.)
https://youtu.be/eJq7L3kO2Ng?t=15820 (He gave up.)
https://youtu.be/eJq7L3kO2Ng?t=16097 (He gave up.)
https://youtu.be/eJq7L3kO2Ng?t=16242 (Another try.)
https://youtu.be/eJq7L3kO2Ng?t=16291 (Another try.)
https://youtu.be/eJq7L3kO2Ng?t=16477 (Another try.)
https://youtu.be/eJq7L3kO2Ng?t=16555 (Another try.)
https://youtu.be/eJq7L3kO2Ng?t=16748 (Another try.)
https://youtu.be/eJq7L3kO2Ng?t=16891 (Another try.)

**Edit.**
Related discussions: https://github.com/ihhub/fheroes2/discussions/4396, https://github.com/ihhub/fheroes2/discussions/4327, https://github.com/ihhub/fheroes2/discussions/3742.

Bad AI combat movement.
Ogre Lords move away from the other troops by attacking the troop of Sprites.
https://youtu.be/eJq7L3kO2Ng?t=2514

Wrong position of the destination Stone Liths during AI turn (https://github.com/ihhub/fheroes2/issues/4141).
https://youtu.be/eJq7L3kO2Ng?t=2904

Bad AI combat attack.
Ghosts should attack Gargoyles rather than Red Dragons.
https://youtu.be/eJq7L3kO2Ng?t=5621

How annoying it is to adjust the position of the adventure map to the puzzle position of the Ultimate Artifact (https://github.com/ihhub/fheroes2/issues/1985).
https://youtu.be/eJq7L3kO2Ng?t=7366
https://youtu.be/eJq7L3kO2Ng?t=7669

The experienced player, used to cast Dimension Door as long as the hero still has some movement points (done, https://github.com/ihhub/fheroes2/pull/5314).
https://youtu.be/eJq7L3kO2Ng?t=9053
https://youtu.be/eJq7L3kO2Ng?t=9525
https://youtu.be/eJq7L3kO2Ng?t=10220
https://youtu.be/eJq7L3kO2Ng?t=10449

The experienced player, used to cast Town Portal as long as the hero still has some movement points (done, https://github.com/ihhub/fheroes2/pull/5314).
https://youtu.be/eJq7L3kO2Ng?t=24597
https://youtu.be/eJq7L3kO2Ng?t=25363

Bad AI combat ranged attack.
AI Liches could shoot in such a way as to avoid shooting at his friendly troop.
https://youtu.be/eJq7L3kO2Ng?t=10154

AI troops do not take into account the dragon attack on two hexagons to position themselves.
https://youtu.be/eJq7L3kO2Ng?t=10588
https://youtu.be/eJq7L3kO2Ng?t=10617
https://youtu.be/eJq7L3kO2Ng?t=10795
https://youtu.be/eJq7L3kO2Ng?t=11313
https://youtu.be/eJq7L3kO2Ng?t=11362
https://youtu.be/eJq7L3kO2Ng?t=11536
https://youtu.be/eJq7L3kO2Ng?t=11575

Bad AI combat ranged attack.
Archers shoot at Bone Dragons when there is no chance to eliminate one, they should shoot at Skeletons.
https://youtu.be/eJq7L3kO2Ng?t=11278

AI Barbarian hero casts Lightning Bolt on Genie without eliminating him before fleeing, whereas the spell could have been cast on Rangers.
https://youtu.be/eJq7L3kO2Ng?t=20465

Wrong position of the destination whirlpool during AI turn (https://github.com/ihhub/fheroes2/issues/5146).
https://youtu.be/eJq7L3kO2Ng?t=21863
https://youtu.be/eJq7L3kO2Ng?t=22050
https://youtu.be/eJq7L3kO2Ng?t=22647

**Comment by oleg-derevenetz** [comment]
Amazing! Not a single crash for more than seven hours.

**Reply by zenseii** [reply]
> https://youtu.be/eJq7L3kO2Ng?t=15762 (One more try.) https://youtu.be/eJq7L3kO2Ng?t=15810 (He gave up.) 


Watching some of them in slow motion like the one above, it seems obvious that what he wants to do is a swap and not a move, which I've been able to address. But yeah all of the other ones should be ironed out when I finally manage to wrap that PR up...

> Amazing! Not a single crash for more than seven hours.

@oleg-derevenetz Why are you so surprised considering how much work you've done on polishing it? Or are you talking about Windows?🤣

**Reply by oleg-derevenetz** [reply]
@zenseii

> Watching some of them in slow motion like the one above, it seems obvious that what he wants to do is a swap and not a move, which I've been able to address.

That's because that's how the chains of heroes are built by the pro. All heroes in the chain have one relatively fast unit, so when previous hero in the chain meets the next hero, the next hero gets full army (and usually all artifacts as well), and the previous hero gets that relatively fast unit of the next hero. Then the hero who received full army and artifacts runs to the next hero in the chain and everything repeats. In fact, the swap is the most frequently performed operation in chains. HoMM3 even has the special buttons to swap troops and artifacts.

> Why are you so surprised considering how much work you've done on polishing it?

Just kidding :)

**Comment by LeHerosInconnu** [comment]
These remarks are too old to be considered.
I close the discussion.
